otherlibs/unix/socketaddr.c

```diff
@@ -117,7 +117,7 @@ value caml_unix_alloc_sockaddr(union sock_addr_union * adr /*in*/,
   if (adr_len < offsetof(struct sockaddr, sa_data)) {
     // Only possible for an unnamed AF_UNIX socket, in
     // which case sa_family might be uninitialized.
-    return alloc_unix_sockaddr(caml_alloc_string(0));
+    CAMLreturn(alloc_unix_sockaddr(caml_alloc_string(0)));
```

Comment:
Actually, looking at this code again, I wonder whether `alloc_unix_sockaddr(caml_alloc_string(0))` is correct. This smells bad because `caml_alloc_string` would in general return a GC-ed value, which is not registered with the GC, and may thus be invalidated by the following allocation in `alloc_unix_sockaddr`. But then `caml_alloc_string(0)` returns an empty string and it may be that those are cached and do not need to be registered, so the code is probably slightly confusing to non-experts but correct (I'm not fully sure).

Then there is a similar issue down below (line 142), with a `caml_alloc_string` that looks non-empty. That looks even less correct to me:

https://github.com/ocaml/ocaml/blob/e9625983e8c3837a183f2c9b0413eca14ee5da0b/otherlibs/unix/socketaddr.c#L142C1-L144C9

I would propose to play it safe rather than subtle, and rewrite both places as follows:

```c
a = caml_alloc_string(...);
res = alloc_unix_sockaddr(a);
```

(in the first case you can then just `CAMLreturn(res);`)

Does that make sense? Apologies for getting you in trickier waters than I initially thought.

Reply:
> This smells bad because caml_alloc_string would in general return a GC-ed value, which is not registered with the GC, and may thus be invalidated by the following allocation in alloc_unix_sockaddr

`alloc_unix_sockaddr` is careful to register its argument as a root during its allocations.  What can possibly go wrong?  Your "play it safe" rewrite makes no difference, I believe.

Reply:
Ah, right: having exactly one unnamed argument is safe, because even if the temporary value is invalidated by the GC, it is not reused by the rest of the caller code. (The same reasoning does not work with two arguments, where one may invalidate the other before the function is called.) I would still prefer to have simple and clearly-safe code -- safety here makes no performance difference.